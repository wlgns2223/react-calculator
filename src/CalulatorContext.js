import React, { useContext, useReducer, createContext } from "react";
import {
  binaryHandler,
  calculateHanlder,
  keyHandler,
  unaryHanlder,
} from "./handler";

export const initialHeaderState = {
  rvalue: null,
  lvalue: null,
  operator: "",
  current: 0,
  keyInput: false,
};

function headerReducer(state, action) {
  switch (action.type) {
    case "key":
      return keyHandler(state, action);
    case "unary":
      return unaryHanlder(state, action);
    case "binary":
      return binaryHandler(state, action);
    case "calculate":
      return calculateHanlder(state, action);
    default:
      throw new Error(`Unhandled Type Error ${action.type}`);
  }
}

const CalculatorStateContext = createContext();
const CalculatorDispatchContext = createContext();

export function CalculatorContext({ children }) {
  const [state, dispatch] = useReducer(headerReducer, initialHeaderState);
  return (
    <CalculatorStateContext.Provider value={state}>
      <CalculatorDispatchContext.Provider value={dispatch}>
        {children}
      </CalculatorDispatchContext.Provider>
    </CalculatorStateContext.Provider>
  );
}

export function useStateContext() {
  const context = useContext(CalculatorStateContext);
  if (!context) {
    throw new Error(`State Context Undefined ${context}`);
  }
  return context;
}

export function useDispatchContext() {
  const context = useContext(CalculatorDispatchContext);
  if (!context) {
    throw new Error(`Dispatch Context Undefined ${context}`);
  }

  return context;
}
