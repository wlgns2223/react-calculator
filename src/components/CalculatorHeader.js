import React from "react";
import styled from "styled-components";

export default function CalculatorHeader({ current }) {
  return (
    <CalculatorHeaderBlock>
      <CalulatorScreen>{current}</CalulatorScreen>
    </CalculatorHeaderBlock>
  );
}

const CalculatorHeaderBlock = styled.div`
  padding: 1rem 0.85rem;
  border-bottom: 1px solid #ffd523;
  width: 100%;
`;

const CalulatorScreen = styled.div`
  background-color: white;
  border: 1px solid white;
  border-radius: 4px;
  width: 100%;
  height: 4rem;
  line-height: 4rem;
  display: flex;
  align-items: center;
  justify-content: flex-end;
  font-size: 3rem;
  overflow-x: scroll;

  @media ${({ theme }) => theme.mediaQueries.medium} {
    height: 5rem;
    font-size: 4rem;
    line-height: 5rem;
  }

  @media ${({ theme }) => theme.mediaQueries.large} {
    height: 6rem;
    line-height: 6rem;
    font-size: 5rem;
  }
`;
