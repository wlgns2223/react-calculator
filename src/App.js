import "./App.css";
import { createGlobalStyle } from "styled-components";
import CalculatorTemplate from "./components/CalculatorTemplate";
import CalculatorHeader from "./components/CalculatorHeader";
import CalculatorKeypad from "./components/CalculatorKeypad";
import { CalculatorContext } from "./CalculatorContext";

const GlobalStyle = createGlobalStyle`
body{
  background-color: white;
}
`;

function App() {
  return (
    <CalculatorContext>
      <GlobalStyle />
      <CalculatorTemplate>
        <CalculatorHeader />
        <CalculatorKeypad />
      </CalculatorTemplate>
    </CalculatorContext>
  );
}

export default App;
