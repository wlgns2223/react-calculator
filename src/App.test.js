import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import App from "./App";
import CalculatorHeader from "./components/CalculatorHeader";
import { CalculatorContext } from "./CalulatorContext";
import { useContext } from "react";

jest.mock("react");

test("<header />", () => {
  useContext.mockImplementation(() => {
    return {
      rvalue: null,
      lvalue: null,
      operator: "",
      current: 0,
      keyInput: false,
    };
  });
  const { getByText } = render(
    // <CalculatorContext>
    <CalculatorHeader />
    // </CalculatorContext>
  );

  const elem = getByText("0");
  expect(elem).toBeInTheDocument();
});
