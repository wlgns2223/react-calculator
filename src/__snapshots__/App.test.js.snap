// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-bdnxRM bNYgnN"
      >
        <div
          class="sc-gtsrHT liWHHl"
        >
          <div
            class="sc-dlnjwi ffTWPa"
          >
            0
          </div>
        </div>
        <div
          class="sc-eCApnc diESMJ"
        >
          <button
            class="sc-hKFxyN kSWias"
            color="darkGray"
          >
            AC
          </button>
          <button
            class="sc-hKFxyN kSWias"
            color="darkGray"
          >
            +/-
          </button>
          <button
            class="sc-hKFxyN kSWias"
            color="darkGray"
          >
            %
          </button>
          <button
            class="sc-hKFxyN ekGPPR"
            color="gold"
          >
            /
          </button>
          <button
            class="sc-hKFxyN eZTReh"
            color="gray"
          >
            7
          </button>
          <button
            class="sc-hKFxyN eZTReh"
            color="gray"
          >
            8
          </button>
          <button
            class="sc-hKFxyN eZTReh"
            color="gray"
          >
            9
          </button>
          <button
            class="sc-hKFxyN ekGPPR"
            color="gold"
          >
            X
          </button>
          <button
            class="sc-hKFxyN eZTReh"
            color="gray"
          >
            4
          </button>
          <button
            class="sc-hKFxyN eZTReh"
            color="gray"
          >
            5
          </button>
          <button
            class="sc-hKFxyN eZTReh"
            color="gray"
          >
            6
          </button>
          <button
            class="sc-hKFxyN ekGPPR"
            color="gold"
          >
            -
          </button>
          <button
            class="sc-hKFxyN eZTReh"
            color="gray"
          >
            1
          </button>
          <button
            class="sc-hKFxyN eZTReh"
            color="gray"
          >
            2
          </button>
          <button
            class="sc-hKFxyN eZTReh"
            color="gray"
          >
            3
          </button>
          <button
            class="sc-hKFxyN ekGPPR"
            color="gold"
          >
            +
          </button>
          <button
            class="sc-hKFxyN isENax"
            color="gray"
          >
            0
          </button>
          <button
            class="sc-hKFxyN eZTReh"
            color="gray"
          >
            .
          </button>
          <button
            class="sc-hKFxyN ekGPPR"
            color="gold"
          >
            =
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-bdnxRM bNYgnN"
    >
      <div
        class="sc-gtsrHT liWHHl"
      >
        <div
          class="sc-dlnjwi ffTWPa"
        >
          0
        </div>
      </div>
      <div
        class="sc-eCApnc diESMJ"
      >
        <button
          class="sc-hKFxyN kSWias"
          color="darkGray"
        >
          AC
        </button>
        <button
          class="sc-hKFxyN kSWias"
          color="darkGray"
        >
          +/-
        </button>
        <button
          class="sc-hKFxyN kSWias"
          color="darkGray"
        >
          %
        </button>
        <button
          class="sc-hKFxyN ekGPPR"
          color="gold"
        >
          /
        </button>
        <button
          class="sc-hKFxyN eZTReh"
          color="gray"
        >
          7
        </button>
        <button
          class="sc-hKFxyN eZTReh"
          color="gray"
        >
          8
        </button>
        <button
          class="sc-hKFxyN eZTReh"
          color="gray"
        >
          9
        </button>
        <button
          class="sc-hKFxyN ekGPPR"
          color="gold"
        >
          X
        </button>
        <button
          class="sc-hKFxyN eZTReh"
          color="gray"
        >
          4
        </button>
        <button
          class="sc-hKFxyN eZTReh"
          color="gray"
        >
          5
        </button>
        <button
          class="sc-hKFxyN eZTReh"
          color="gray"
        >
          6
        </button>
        <button
          class="sc-hKFxyN ekGPPR"
          color="gold"
        >
          -
        </button>
        <button
          class="sc-hKFxyN eZTReh"
          color="gray"
        >
          1
        </button>
        <button
          class="sc-hKFxyN eZTReh"
          color="gray"
        >
          2
        </button>
        <button
          class="sc-hKFxyN eZTReh"
          color="gray"
        >
          3
        </button>
        <button
          class="sc-hKFxyN ekGPPR"
          color="gold"
        >
          +
        </button>
        <button
          class="sc-hKFxyN isENax"
          color="gray"
        >
          0
        </button>
        <button
          class="sc-hKFxyN eZTReh"
          color="gray"
        >
          .
        </button>
        <button
          class="sc-hKFxyN ekGPPR"
          color="gold"
        >
          =
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
